pallets/liquidity-pools-gateway/src/message_processing.rs

```diff
@@ -0,0 +1,489 @@
+use cfg_primitives::LP_DEFENSIVE_WEIGHT;
+use cfg_traits::liquidity_pools::{
+	InboundMessageHandler, LPEncoding, MessageQueue, MessageSender, Proof, RouterProvider,
+};
+use cfg_types::domain_address::{Domain, DomainAddress};
+use frame_support::{
+	dispatch::DispatchResult,
+	ensure,
+	pallet_prelude::{Decode, Encode, Get, TypeInfo},
+	weights::Weight,
+};
+use parity_scale_codec::MaxEncodedLen;
+use sp_arithmetic::traits::{EnsureAddAssign, EnsureSub};
+use sp_runtime::DispatchError;
+
+use crate::{
+	message::GatewayMessage, Config, Error, Pallet, PendingInboundEntries, Routers, SessionIdStore,
+};
+
+/// Type used when storing inbound message information.
+#[derive(Debug, Encode, Decode, Clone, Eq, MaxEncodedLen, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub enum InboundEntry<T: Config> {
+	Message {
+		session_id: T::SessionId,
+		domain_address: DomainAddress,
+		message: T::Message,
+		expected_proof_count: u32,
+	},
+	Proof {
+		session_id: T::SessionId,
+		current_count: u32,
+	},
+}
+
+/// Type used when processing inbound messages.
+#[derive(Clone)]
+pub struct InboundProcessingInfo<T: Config> {
+	pub domain_address: DomainAddress,
+	pub router_ids: Vec<T::RouterId>,
+	pub current_session_id: T::SessionId,
+	pub expected_proof_count_per_message: u32,
+}
+
+impl<T: Config> Pallet<T> {
+	/// Retrieves all available routers for a domain and then filters them based
+	/// on the routers that we have in storage.
+	pub fn get_router_ids_for_domain(domain: Domain) -> Result<Vec<T::RouterId>, DispatchError> {
+		let all_routers_for_domain = T::RouterProvider::routers_for_domain(domain);
+
+		let stored_routers = Routers::<T>::get();
+
+		let res = all_routers_for_domain
+			.iter()
+			.filter(|x| stored_routers.iter().any(|y| *x == y))
+			.map(|x| x.clone())
+			.collect::<Vec<_>>();
+
+		if res.is_empty() {
+			return Err(Error::<T>::NotEnoughRoutersForDomain.into());
+		}
+
+		Ok(res)
+	}
```

Comment:
We should invert this @cdamian, if configuration list priority matter, then we should iterate `stored_routers` first and filter with `router_for_domain`

Reply:
Awesome catch!

Reply:
Coming back to:

>Suppose gateway is configured with the following (dynamically configured by an operator):
AxelarXcm
AxelarEvm
SnowbridgeEvm
>
>And the chain offers the following routers (statically configured in the runtime):
AxelarXcm
AxelarEvm
SnowbridgeEvm
OtherEvm
>
>If a message with Domain::Evm needs to be sent, it must be sent only to AxelarEvm and SnowbridgeEvm.
AxelarXcm will not used because that router is not retrieved by for_domain(Domain::Evm)
OtherEvm will not used because it's not configured in the gateway

The operator now has to ensure the correct order of routers while also keeping filtering in mind.

Reply:
Agree! It's a ordered whitelist of routers